File: strings/java/Solution06.java
package strings.java;

public class Solution06 {

    public int myAtoi(String s) {

        if (s == null || s.isEmpty()) {
            return 0;
        }

        int i = 0;
        int n = s.length();

        while (i < n && s.charAt(i) == ' ') {
            i++;
        }

        int sign = 1;

        if (i < n) {
            if (s.charAt(i) == '-') {
                sign = -1;
                i++;
            } else if (s.charAt(i) == '+') {
                i++;
            }
        }

        long result = 0;

        while (i < n && Character.isDigit(s.charAt(i))) {
            int digit = s.charAt(i) - '0';
            result = result * 10 + digit;

            if (sign == 1 && result > Integer.MAX_VALUE) {
                return Integer.MAX_VALUE;
            }

            if (sign == -1 && -result < Integer.MIN_VALUE) {
                return Integer.MIN_VALUE;
            }

            i++;
        }
        return (int) (sign * result);
    }
}

File: strings/go/solution06.go
package leetcode

import (
	"math"
	"unicode"
)

func myAtoi(s string) int {

	i := 0
	n := len(s)

	for i < n && s[i] == ' ' {
		i++
	}

	sign := 1

	if i < n {
		switch s[i] {
		case '-':
			sign = -1
			i++
		case '+':
			i++
		}
	}

	var result int64 = 0

	for i < n && unicode.IsDigit(rune(s[i])) {
		digit := s[i] - '0'
		result = result*10 + int64(digit)

		if sign == 1 && result > math.MaxInt32 {
			return math.MaxInt32
		}

		if sign == -1 && result < math.MinInt32 {
			return math.MinInt32
		}

		i++
	}

	return int(sign * int(result))
}
